eth_defi/uniswap_v3/price.py

```diff
@@ -81,3 +84,226 @@ def validate_args(path, fees, slippage, amount):
+        amount_eth = estimate_buy_quantity(
+            uniswap_v3,
+            weth.address,
+            usdc.address,
+            500*10**18,
+            500,
+        )
+        assert amount_eth / 1e18 == pytest.approx(0.28488156127668085)
+
+    :param quantity: How much of the quote token we have to use
+    :param uniswap: Uniswap v3 deployment
+    :param base_token_address: Address of the base token of the trading pair
+    :param quote_token_address: Address of the quote token of the trading pair
+    :param target_pair_fee: Trading fee express in raw format for the target pair
+    :param slippage: Slippage express in bps
+    :return: Expected base token to receive
+    """
+    price_helper = UniswapV3PriceHelper(uniswap)
+    
+    if intermediate_token_address:
+        path = [quote_token_address, intermediate_token_address, base_token_address]
+        fees = [intermediate_pair_fee, target_pair_fee]
+    else:
+        path = [quote_token_address, base_token_address]
+        fees = [target_pair_fee]
+
+    return price_helper.get_amount_out(quantity, path, fees, slippage=slippage)
+
+
+def estimate_buy_price(
+    uniswap: UniswapV3Deployment,
+    base_token_address: HexAddress,
+    quote_token_address: HexAddress,
+    quantity: int,
+    target_pair_fee: int,
+    *,
+    slippage: float = 0,
+    intermediate_token_address: HexAddress | None = None,
+    intermediate_pair_fee: int | None = None,
+) -> int:
+    """Estimate how much we are going to need to pay when doing buy.
+
+    Calls the on-chain contract to get the current liquidity and estimates the
+    the price based on it.
+
+    Example:
+
+    .. code-block:: python
+
+        # Estimate how much ETH we will receive for 500 USDC.
+        # In this case the pool ETH price is $1700 so this should be below ~1/4 of ETH
+        amount_eth = estimate_buy_price(
+            uniswap_v3,
+            weth.address,
+            usdc.address,
+            1*10**18,
+            500,
+        )
+        assert amount_eth / 1e18 == pytest.approx(0.28488156127668085)
+
+    :param uniswap: Uniswap v2 deployment
+    :param base_token_address: Base token address of the trading pair
+    :param quote_token_address: Quote token addressof the trading pair
+    :param quantity: How much of the base token we want to buy
+    :param target_pair_fee: Trading fee express in raw format for the target pair
+    :param slippage: Slippage express in bps
+    :return: Expected base token to receive
+    """
+    price_helper = UniswapV3PriceHelper(uniswap)
+    if intermediate_token_address:
+        path = [quote_token_address, intermediate_token_address, base_token_address]
+        fees = [intermediate_pair_fee, target_pair_fee]
+    else:
+        path = [quote_token_address, base_token_address]
+        fees = [target_pair_fee]
+
+    return price_helper.get_amount_in(quantity, path, fees, slippage=slippage)
+
+
+def estimate_sell_price(
```

Comment:
This can be probably removed as well, since it's almost exactly the same as `estimate_sell_received_amount_raw` below